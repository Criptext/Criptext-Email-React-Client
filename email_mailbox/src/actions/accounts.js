import { Account } from './types';

export const addAccounts = accounts => {
  return {
    type: Account.ADD_BATCH,
    accounts
  };
};
