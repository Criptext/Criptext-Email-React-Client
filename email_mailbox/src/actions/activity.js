import { Activity } from './types';

export const logout = () => {
  return {
    type: Activity.LOGOUT
  };
};

export const stopAll = () => {
  return {
    type: Activity.STOP_ALL
  };
};

export const startLoadSync = () => {
  return {
    type: Activity.START_LOAD_SYNC
  };
};

export const stopLoadSync = () => {
  return {
    type: Activity.STOP_LOAD_SYNC
  };
};

export const startLoadThread = () => {
  return {
    type: Activity.START_LOAD_THREAD
  };
};

export const stopLoadThread = () => {
  return {
    type: Activity.STOP_LOAD_THREAD
  };
};

export const setAvatarUpdatedTimestamp = timestamp => {
  return {
    type: Activity.AVATAR_UPDATED_TIMESTAMP,
    timestamp
  };
};

export const updateLoadingSync = ({ totalTask, completedTask }) => {
  return {
    type: Activity.UPDATE_LOADING_SYNC,
    totalTask,
    completedTask
  };
};

export const updateSwitchThreads = ({ checked, disabled }) => {
  return {
    type: Activity.UPDATE_SWITCH_THREADS,
    checked,
    disabled
  };
};
