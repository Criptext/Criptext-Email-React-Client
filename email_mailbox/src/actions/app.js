import { App } from './types';
import {
  addAccounts,
  addContacts,
  addFeedItems,
  addLabels,
  addThreads,
  stopLoadThread
} from './index';
import { assembleAccounts } from './../utils/AccountUtils';
import { assembleLabels } from './../utils/LabelUtils';
import { assembleThreads } from './../utils/ThreadUtils';
import { assembleFeedItems } from './../utils/FeedItemUtils';
import { getGroupEvents } from './../utils/electronEventInterface';
const INIT_LIMIT_THREADS = 22;

export const addDataApp = ({
  account,
  activity,
  contact,
  email,
  file,
  feeditem,
  label,
  thread
}) => ({
  type: App.ADD_DATA,
  account,
  activity,
  contact,
  email,
  file,
  feeditem,
  label,
  thread
});

export const loadApp = params => {
  return async dispatch => {
    const accounts = await assembleAccounts();
    const labels = await assembleLabels();
    const { threads, contacts } = await assembleThreads({
      ...params,
      limit: INIT_LIMIT_THREADS
    });
    const feeditems = await assembleFeedItems();

    const account = addAccounts(accounts);
    const activity = stopLoadThread();
    const contact = addContacts(contacts);
    const feeditem = addFeedItems(feeditems, true);
    const label = addLabels(labels);
    const thread = addThreads(params.labelId, threads, true);

    dispatch(
      addDataApp({ account, activity, contact, label, feeditem, thread })
    );
    await getGroupEvents({});
  };
};
