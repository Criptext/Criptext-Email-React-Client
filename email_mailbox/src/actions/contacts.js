import { Contact } from './types';
import { defineContacts } from './../utils/ContactUtils';

export const addContacts = contacts => {
  return {
    type: Contact.ADD_BATCH,
    contacts
  };
};

export const loadContacts = ids => {
  return async dispatch => {
    try {
      const contacts = await defineContacts(ids);
      dispatch(addContacts(contacts));
    } catch (e) {
      // TO DO
    }
  };
};
