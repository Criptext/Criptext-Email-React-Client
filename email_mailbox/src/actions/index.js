import { addAccounts } from './accounts';
import {
  logout,
  setAvatarUpdatedTimestamp,
  startLoadSync,
  startLoadThread,
  stopAll,
  stopLoadSync,
  stopLoadThread,
  updateLoadingSync,
  updateSwitchThreads
} from './activity';
import { addContacts, loadContacts } from './contacts';
import { addFiles, loadFiles, unsendEmailFiles } from './files';
import {
  addLabelIdThread,
  addLabelIdThreadSuccess,
  addLabelIdThreadDraft,
  addLabelIdThreads,
  addLabelIdThreadsSuccess,
  addMoveLabelIdThreads,
  addThreads,
  filterThreadsOrLoadMoreByUnread,
  loadEvents,
  loadThreads,
  moveThreads,
  removeLabelIdThread,
  removeLabelIdThreadSuccess,
  removeLabelIdThreadDraft,
  removeLabelIdThreads,
  removeLabelIdThreadsSuccess,
  updateEmailIdsThread,
  updateThreadsSuccess,
  updateUnreadThreads,
  removeThreads,
  removeThreadsDrafts,
  removeThreadsSuccess,
  sendOpenEvent,
  updateThread
} from './threads';
import {
  addEmails,
  loadEmails,
  markEmailUnread,
  muteEmail,
  muteNotifications,
  removeEmails,
  removeEmailsOnSuccess,
  unsendEmail,
  unsendEmailOnSuccess,
  updateEmailLabels,
  updateEmailOnSuccess
} from './emails';
import {
  addLabels,
  addLabel,
  removeLabel,
  removeLabelOnSuccess,
  updateBadgeLabels,
  updateLabel,
  updateLabelSuccess
} from './labels';
import {
  addFeedItems,
  loadFeedItems,
  removeFeedItem,
  removeFeedItemSuccess,
  selectFeedItem,
  updateAllFeedItemsAsOlder,
  updateFeedItemSuccess
} from './feeditems';
import { loadSuggestions } from './suggestions';
import { addDataApp, loadApp } from './app';

export {
  addAccounts,
  addContacts,
  addDataApp,
  addEmails,
  addFiles,
  addFeedItems,
  addLabel,
  addLabels,
  addLabelIdThread,
  addLabelIdThreadSuccess,
  addLabelIdThreadDraft,
  addLabelIdThreads,
  addLabelIdThreadsSuccess,
  addMoveLabelIdThreads,
  addThreads,
  filterThreadsOrLoadMoreByUnread,
  loadApp,
  loadContacts,
  loadEmails,
  loadEvents,
  loadFiles,
  loadFeedItems,
  loadSuggestions,
  loadThreads,
  logout,
  markEmailUnread,
  moveThreads,
  muteEmail,
  muteNotifications,
  removeEmails,
  removeEmailsOnSuccess,
  removeLabel,
  removeLabelIdThread,
  removeLabelIdThreadSuccess,
  removeLabelIdThreadDraft,
  removeLabelIdThreads,
  removeLabelIdThreadsSuccess,
  removeLabelOnSuccess,
  removeFeedItem,
  removeFeedItemSuccess,
  removeThreads,
  removeThreadsDrafts,
  removeThreadsSuccess,
  selectFeedItem,
  sendOpenEvent,
  setAvatarUpdatedTimestamp,
  startLoadSync,
  startLoadThread,
  stopAll,
  stopLoadSync,
  stopLoadThread,
  unsendEmail,
  unsendEmailFiles,
  unsendEmailOnSuccess,
  updateAllFeedItemsAsOlder,
  updateBadgeLabels,
  updateFeedItemSuccess,
  updateEmailIdsThread,
  updateEmailLabels,
  updateEmailOnSuccess,
  updateLabel,
  updateLabelSuccess,
  updateLoadingSync,
  updateSwitchThreads,
  updateThread,
  updateThreadsSuccess,
  updateUnreadThreads
};
