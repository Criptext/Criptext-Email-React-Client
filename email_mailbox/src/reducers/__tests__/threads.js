/* eslint-env node, jest */

import threadsReducer from './../threads';
import * as actions from './../../actions/index';
import file from './../../../public/threads.json';

jest.mock('./../../utils/const');
jest.mock('./../../utils/ipc');
jest.mock('./../../utils/electronInterface');
jest.mock('./../../utils/electronEventInterface');

const myThreads = file.threads;

function initState(labelId, threads) {
  return threadsReducer(undefined, actions.addThreads(labelId, threads));
}

describe('Thread actions - ADD_BATCH', () => {
  const manyThreads = [myThreads[0], myThreads[1]];
  const labelId = 1;

  it('should add threads to state', () => {
    expect(initState(labelId, manyThreads)).toMatchSnapshot();
  });
});

describe('Thread actions - ADD_LABELID_THREAD', () => {
  const threads = [myThreads[0]];

  it('should update thread params: allLabels and labels', () => {
    const labelId = 1;
    const state = initState(labelId, threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const labelIdToAdd = 5;
    const action = actions.addLabelIdThreadSuccess(
      labelId,
      threadId,
      labelIdToAdd
    );
    const newState = threadsReducer(state, action);
    const threadUpdated = newState
      .get(`${labelId}`)
      .get('list')
      .get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 5],
        labels: [2, 3, 5]
      })
    );
    const allIds = newState.get(`${labelId}`).get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });

  it('should not update thread param: allLabels and labels, when threadId is undefined', () => {
    const labelId = 1;
    const state = initState(labelId, threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const labelIdToAdd = undefined;
    const action = actions.addLabelIdThreadSuccess(
      labelId,
      threadId,
      labelIdToAdd
    );
    const newState = threadsReducer(state, action);
    const threadUpdated = newState
      .get(`${labelId}`)
      .get('list')
      .get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3],
        labels: [2, 3]
      })
    );
    const allIds = newState.get(`${labelId}`).get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });

  it('should not update thread param: allLabels and labels, when labelId is not number type', () => {
    const labelId = 1;
    const state = initState(labelId, threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const labelIdToAdd = '4';
    const action = actions.addLabelIdThreadSuccess(
      labelId,
      threadId,
      labelIdToAdd
    );
    const newState = threadsReducer(state, action);
    const threadUpdated = newState
      .get(`${labelId}`)
      .get('list')
      .get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3],
        labels: [2, 3]
      })
    );
    const allIds = newState.get(`${labelId}`).get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });
});

describe('Thread actions - ADD_LABELID_THREAD_DRAFT', () => {
  const threads = [myThreads[1]];

  it('should update thread params: allLabels and labels', () => {
    const state = initState(threads);
    const uniqueId = 2;
    const labelId = 10;
    const action = actions.addLabelIdThreadDraftSuccess(uniqueId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [5, 10],
        labels: [10]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([2]);
  });

  it('should not update thread param: allLabels and labels, when uniqueId is not number type', () => {
    const state = initState(threads);
    const uniqueId = '2';
    const labelId = undefined;
    const action = actions.addLabelIdThreadDraftSuccess(uniqueId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [5],
        labels: []
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([2]);
  });

  it('should not update thread param: allLabels and labels, when labelId is not number type', () => {
    const state = initState(threads);
    const uniqueId = 2;
    const labelId = '4';
    const action = actions.addLabelIdThreadDraftSuccess(uniqueId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [5],
        labels: []
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([2]);
  });
});

describe('Thread actions - ADD_LABELID_THREADS', () => {
  const threads = [myThreads[0], myThreads[2]];

  it('should update threads params: allLabels and labels', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh'];
    const labelId = 5;
    const action = actions.addLabelIdThreadsSuccess(threadIds, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const threadUpdated2 = newState.get('list').get('1');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4, 5],
        labels: [1, 2, 3, 5]
      })
    );
    expect(threadUpdated2.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 5],
        labels: [1, 2, 5]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(threadIds);
  });

  it('should not update threads param: allLabels and labels, when threadIds is undefined', () => {
    const state = initState(threads);
    const threadIds = undefined;
    const labelId = 5;
    const action = actions.addLabelIdThreadsSuccess(threadIds, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const threadUpdated2 = newState.get('list').get('1');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4],
        labels: [1, 2, 3]
      })
    );
    expect(threadUpdated2.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3],
        labels: [1, 2]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh']);
  });

  it('should not update threads param: allLabels and labels, when labelId is not number type', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh'];
    const labelId = '5';
    const action = actions.addLabelIdThreadsSuccess(threadIds, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const threadUpdated2 = newState.get('list').get('1');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4],
        labels: [1, 2, 3]
      })
    );
    expect(threadUpdated2.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3],
        labels: [1, 2]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(threadIds);
  });
});

describe('Thread actions - MOVE_THREADS', () => {
  const threads = [myThreads[0], myThreads[2]];

  it('should move threads params: threadIds', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9'];
    const action = actions.moveThreads(threadIds);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list');
    expect(threadUpdated.size).toEqual(1);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['cuW6ElyoqsR7MMh']);
  });

  it('should move threads params: threadIds not exist', () => {
    const state = initState(threads);
    const threadIds = null;
    const action = actions.moveThreads(threadIds);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list');
    expect(threadUpdated.size).toEqual(2);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh']);
  });

  it('should move threads params: threadIds is different', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dMlE0OSS9'];
    const action = actions.moveThreads(threadIds);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list');
    expect(threadUpdated.size).toEqual(2);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh']);
  });
});

describe('Thread actions - REMOVE_LABELID_THREAD', () => {
  const threads = [myThreads[0]];

  it('should update thread params: allLabels and labels', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const labelId = 1;
    const action = actions.removeLabelIdThreadSuccess(threadId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [4, 2, 3],
        labels: [3, 2]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });

  it('should not update thread param: allLabels and labels, when threadId is undefined', () => {
    const state = initState(threads);
    const threadId = undefined;
    const labelId = 1;
    const action = actions.removeLabelIdThreadSuccess(threadId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4],
        labels: [1, 2, 3]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });

  it('should not update thread param: allLabels and labels, when labelId is not number type', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const labelId = '1';
    const action = actions.removeLabelIdThreadSuccess(threadId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4],
        labels: [1, 2, 3]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });
});

describe('Thread actions - REMOVE_LABELID_THREAD_DRAFT', () => {
  const threads = [myThreads[1]];

  it('should update thread params: allLabels and labels', () => {
    const state = initState(threads);
    const uniqueId = 2;
    const labelId = 5;
    const action = actions.removeLabelIdThreadDraftSuccess(uniqueId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [],
        labels: []
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([2]);
  });

  it('should not update thread param: allLabels and labels, when uniqueId is not number type', () => {
    const state = initState(threads);
    const uniqueId = '2';
    const labelId = undefined;
    const action = actions.removeLabelIdThreadDraftSuccess(uniqueId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [5],
        labels: []
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([2]);
  });

  it('should not update thread param: allLabels and labels, when labelId is not number type', () => {
    const state = initState(threads);
    const uniqueId = 2;
    const labelId = '5';
    const action = actions.removeLabelIdThreadDraftSuccess(uniqueId, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [5],
        labels: []
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([2]);
  });
});

describe('Thread actions - REMOVE_LABELID_THREADS', () => {
  const threads = [myThreads[0], myThreads[2]];

  it('should update threads params: allLabels and labels', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh'];
    const labelId = 1;
    const action = actions.removeLabelIdThreadsSuccess(threadIds, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const threadUpdated2 = newState.get('list').get('1');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [4, 2, 3],
        labels: [3, 2]
      })
    );
    expect(threadUpdated2.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [3, 2],
        labels: [2]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(threadIds);
  });

  it('should not update threads param: allLabels and labels, when threadIds is undefined', () => {
    const state = initState(threads);
    const threadIds = undefined;
    const labelId = 1;
    const action = actions.removeLabelIdThreadsSuccess(threadIds, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const threadUpdated2 = newState.get('list').get('1');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4],
        labels: [1, 2, 3]
      })
    );
    expect(threadUpdated2.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3],
        labels: [1, 2]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh']);
  });

  it('should not update threads param: allLabels and labels, when labelId is not number type', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh'];
    const labelId = '1';
    const action = actions.removeLabelIdThreadsSuccess(threadIds, labelId);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const threadUpdated2 = newState.get('list').get('1');
    expect(threadUpdated.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3, 4],
        labels: [1, 2, 3]
      })
    );
    expect(threadUpdated2.toJS()).toMatchObject(
      expect.objectContaining({
        allLabels: [1, 2, 3],
        labels: [1, 2]
      })
    );
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(threadIds);
  });
});

describe('Thread actions - REMOVE_THREADS', () => {
  const threads = [myThreads[0], myThreads[2]];

  it('should remove threads params: threadIds', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9'];
    const action = actions.removeThreadsSuccess(threadIds);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list');
    expect(threadUpdated.size).toEqual(1);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['cuW6ElyoqsR7MMh']);
  });

  it('should move threads params: threadIds not exist', () => {
    const state = initState(threads);
    const threadIds = null;
    const action = actions.removeThreadsSuccess(threadIds);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list');
    expect(threadUpdated.size).toEqual(2);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh']);
  });

  it('should move threads params: threadIds is different', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dMlE0OSS9'];
    const action = actions.removeThreadsSuccess(threadIds);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list');
    expect(threadUpdated.size).toEqual(2);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9', 'cuW6ElyoqsR7MMh']);
  });
});

describe('Thread actions - UPDATE_EMAILIDS_THREAD', () => {
  const threads = [myThreads[0]];

  it('should update thread param: emailIds, add and remove', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const emailIdToAdd = 4;
    const emailIdsToRemove = [1];
    const action = actions.updateEmailIdsThread({
      threadId,
      emailIdToAdd,
      emailIdsToRemove
    });
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const emailIds = threadUpdated.get('emailIds').toJS();
    expect(emailIds).toEqual([2, 4]);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([threadId]);
  });

  it('should update thread param: emailIds, just add', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const emailIdToAdd = 4;
    const action = actions.updateEmailIdsThread({ threadId, emailIdToAdd });
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const emailIds = threadUpdated.get('emailIds').toJS();
    expect(emailIds).toEqual([1, 2, 4]);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([threadId]);
  });

  it('should update thread param: emailIds, just remove', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const emailIdsToRemove = [1];
    const action = actions.updateEmailIdsThread({ threadId, emailIdsToRemove });
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const emailIds = threadUpdated.get('emailIds').toJS();
    expect(emailIds).toEqual([2]);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([threadId]);
  });

  it('should update thread param: emailIds', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const emailIds = [100, 200, 300];
    const action = actions.updateEmailIdsThread({ threadId, emailIds });
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const emailIdsUpdated = threadUpdated.get('emailIds').toJS();
    expect(emailIdsUpdated).toEqual(emailIds);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([threadId]);
  });

  it('should not update thread param: emailIds, when threadId is undefined', () => {
    const state = initState(threads);
    const threadId = undefined;
    const emailIdToAdd = 4;
    const emailIdsToRemove = 1;
    const action = actions.updateEmailIdsThread({
      threadId,
      emailIdToAdd,
      emailIdsToRemove
    });
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const emailIds = threadUpdated.get('emailIds').toJS();
    expect(emailIds).toEqual([1, 2]);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });

  it('should not update thread param: emailIds, when emailIdToAdd and emailIdToRemove are undefined', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const action = actions.updateEmailIdsThread({ threadId });
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const emailIds = threadUpdated.get('emailIds').toJS();
    expect(emailIds).toEqual([1, 2]);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual([threadId]);
  });

  it('should remove thread param: emailIds', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const emailIdsToRemove = [1, 2];
    const action = actions.updateEmailIdsThread({ threadId, emailIdsToRemove });
    const newState = threadsReducer(state, action);
    const listUpdated = newState.get('list');
    expect(listUpdated.size).toEqual(0);
    const allIds = newState.get('allIds');
    expect(allIds.size).toEqual(0);
  });
});

describe('Thread actions - UPDATE_STATUS_THREAD', () => {
  const threads = [myThreads[0]];

  it('should update thread param: status', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const newStatus = 2;
    const action = actions.updateStatusThread(threadId, newStatus);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const status = threadUpdated.get('status');
    expect(status).toBe(newStatus);
  });

  it('should not update thread param: status when status is not number type', () => {
    const state = initState(threads);
    const threadId = '6Za2dcMlE0OSSc9';
    const newStatus = '1';
    const action = actions.updateStatusThread(threadId, newStatus);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const status = threadUpdated.get('status');
    expect(status).not.toBe(newStatus);
    expect(status).toBe(0);
  });
});

describe('Thread actions - UPDATE_UNREAD_THREADS', () => {
  const threads = [myThreads[0]];

  it('should update thread param: unread', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9'];
    const action = actions.updateUnreadThreadsSuccess(threadIds, false);
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const unread = threadUpdated.get('unread');
    expect(unread).toBe(false);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(threadIds);
  });

  it('should not update thread param: unread, when unread is not bool type', () => {
    const state = initState(threads);
    const threadIds = ['6Za2dcMlE0OSSc9'];
    const action = actions.updateUnreadThreadsSuccess(threadIds, 'false');
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const unread = threadUpdated.get('unread');
    expect(unread).toBe(true);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(threadIds);
  });

  it('should not update thread param: unread, when threadIds is empty', () => {
    const state = initState(threads);
    const threadIds = [];
    const action = actions.updateUnreadThreadsSuccess(threadIds, 'true');
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const unread = threadUpdated.get('unread');
    expect(unread).toBe(true);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });

  it('should not update thread param: unread, when threadIds is undefined', () => {
    const state = initState(threads);
    const threadIds = undefined;
    const action = actions.updateUnreadThreadsSuccess(threadIds, 'true');
    const newState = threadsReducer(state, action);
    const threadUpdated = newState.get('list').get('0');
    const unread = threadUpdated.get('unread');
    expect(unread).toBe(true);
    const allIds = newState.get('allIds');
    expect(Array.from(allIds)).toEqual(['6Za2dcMlE0OSSc9']);
  });
});
