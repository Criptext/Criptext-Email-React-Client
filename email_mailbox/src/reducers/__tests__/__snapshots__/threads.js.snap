// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thread actions - ADD_BATCH should add threads to state 1`] = `
Immutable.Map {
  "1": Immutable.Map {
    "list": Immutable.List [
      Immutable.Map {
        "allLabels": Immutable.Set [
          1,
          2,
          3,
        ],
        "fromContactName": Immutable.List [
          "User a",
        ],
        "lastEmailId": "6Za2dcMlE0OSSc9",
        "threadId": "6Za2dcMlE0OSSc9",
        "trashDate": null,
        "preview": "v3m cNvAeKiY1EakWwnfTvLdO22YzkuS7bILjnWncQCwqHBHWfYNsQekd4A3",
        "uniqueId": "6Za2dcMlE0OSSc9",
        "subject": "v3m cNvAeKiY1EakWwnfTvLdO22YzkuS7bILjnWncQCwqHBHWfYNsQekd4A3",
        "date": "2018-09-10 16:13:30",
        "unread": true,
        "recipientContactIds": Immutable.Set [
          1,
        ],
        "status": 0,
        "secure": true,
        "content": "v3m cNvAeKiY1EakWwnfTvLdO22YzkuS7bILjnWncQCwqHBHWfYNsQekd4A3",
        "id": 1,
        "emailIds": Immutable.List [
          1,
          2,
        ],
        "key": "6Za2dcMlE0OSSc9",
      },
      Immutable.Map {
        "allLabels": Immutable.Set [
          5,
        ],
        "fromContactName": Immutable.List [
          "User a",
        ],
        "lastEmailId": "cuW6ElyoqsR7MMw",
        "threadId": null,
        "trashDate": null,
        "preview": "Jq pUgbu1A0gywTYTeKuWfyfCNGgvWKnAK6SHx0ur 7qcMH3PU5S7ESoLTA3",
        "uniqueId": 2,
        "subject": "Jq pUgbu1A0gywTYTeKuWfyfCNGgvWKnAK6SHx0ur 7qcMH3PU5S7ESoLTA3",
        "date": "2018-09-10 16:23:30",
        "unread": true,
        "recipientContactIds": Immutable.Set [
          1,
        ],
        "status": 0,
        "secure": true,
        "content": "Jq pUgbu1A0gywTYTeKuWfyfCNGgvWKnAK6SHx0ur 7qcMH3PU5S7ESoLTA3",
        "id": 2,
        "emailIds": Immutable.List [
          2,
        ],
        "key": "cuW6ElyoqsR7MMw",
      },
    ],
    "allIds": Immutable.Set [
      "6Za2dcMlE0OSSc9",
      2,
    ],
  },
}
`;
